### PR TITLE
feature(xjx): some incompatible refactorings

### PR DESCRIPTION
**Pull request by sailxjx**
In the process of writing league code, some incompatible refactorings have been made, so I hope to merge it into main as soon as possible, so as to prevent future code from calling non-existing interfaces.

Changelog:

1. Merge task.emit and task.emit_remote methods.
2. Support slurm platform in ditask.
3. Move code of middleware in main_league into middleware dir.
4. Change 'task.emit' event to 'task._emit'.
5. Fix some bugs in Parallel.
6. Forbidden syncing context between processes.
7. Support str in match_labels

### CODE REVIEW

File: ding/entry/cli_parsers/slurm_parser.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import os
+import re
+
+
+class SlurmParser():
+
+    def __init__(self, platform_spec, **kwargs) -> None:
```
Comment (PaParaZz1) [inline ding/entry/cli_parsers/slurm_parser.py:7]: add python typing lint

File: ding/entry/cli_parsers/slurm_parser.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        self.nodelist = self._parse_node_list()
+
+    def parse(self):
+        assert len(self.tasks) == self.ntasks
+        procid = int(os.environ["SLURM_PROCID"])
+        nodename = os.environ["SLURMD_NODENAME"]
+        task = self._get_node_args(procid)
+        # Validation
+        assert task["address"] == nodename
+        return {**self.kwargs, **task}
```
Comment (PaParaZz1) [inline ding/entry/cli_parsers/slurm_parser.py:25]: deepcopy self.kwargs for security

File: ding/entry/cli_parsers/tests/test_slurm_parser.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import pytest
+import os
+from ding.entry.cli_parsers import PLATFORM_PARSERS
+slurm_parser = PLATFORM_PARSERS["slurm"]
+
+
+@pytest.fixture
+def set_slurm_env():
+    os.environ["SLURM_NTASKS"] = '6'  # 参数 n，总进程/任务数
```
Comment (PaParaZz1) [inline ding/entry/cli_parsers/tests/test_slurm_parser.py:9]: use english comment

File: ding/entry/cli_parsers/slurm_parser.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    def _parse_node_list(self):
+        nodelist = os.environ["SLURM_NODELIST"]
+        result = re.match(r"(.*)?\[(.*)\]$", nodelist)
+        if result:
+            prefix, tails = result.groups()
+            nodelist = []
+            for tail in tails.split(","):
+                if "-" in tail:
+                    start, stop = tail.split("-")
```
Comment (PaParaZz1) [inline ding/entry/cli_parsers/slurm_parser.py:35]: this branch is not covered by test

File: ding/scripts/main_league_slurm.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+export LC_ALL=en_US.utf-8
+export LANG=en_US.utf-8
+BASEDIR=$(dirname "$0")
+# srun -p Cerebra_Share --quotatype=reserved --mpi=pmi2 -n6 --ntasks-per-node=3 bash ding/scripts/main_league_slurm.sh
```
Comment (PaParaZz1) [inline ding/scripts/main_league_slurm.sh:6]: remove sensntive information

**Review by PaParaZz1 — APPROVED** [review]

File: ding/entry/cli_parsers/slurm_parser.py (hunk 2 above)
Comment (sailxjx) [reply ding/entry/cli_parsers/slurm_parser.py:25]: this expression will not overwrite properties of kwargs, so there is no need to do deepcopy here.

**Pull request merged**